opal/mca/threads/qthreads/threads_qthreads_mutex.c

```diff
@@ -56,36 +58,81 @@ OBJ_CLASS_INSTANCE(opal_recursive_mutex_t,
                    opal_recursive_mutex_construct,
                    opal_mutex_destruct);
 
-static void opal_cond_create(opal_cond_t *cond)
-{
-    ensure_init_qthreads();
-}
-
 int opal_cond_init(opal_cond_t *cond)
 {
-    return OPAL_ERR_NOT_IMPLEMENTED;
+    opal_atomic_lock_init(&cond->m_lock, 0);
+    cond->m_waiter_head = NULL;
+    cond->m_waiter_tail = NULL;
+    return OPAL_SUCCESS;
 }
 
+typedef struct {
+    int m_signaled;
+    void *m_prev;
+} cond_waiter_t;
+
 int opal_cond_wait(opal_cond_t *cond, opal_mutex_t *lock)
 {
-    ensure_init_qthreads();
-    return 0 == qthread_readFE(*cond, &lock->m_lock_qthreads) ? OPAL_SUCCESS
-                                                              : OPAL_ERROR;
+    opal_threads_ensure_init_qthreads();
+    /* This thread is taking "lock", so only this thread can access this
+     * condition variable.  */
+    opal_atomic_lock(&cond->m_lock);
+    cond_waiter_t waiter = { 0, NULL };
+    if (NULL == cond->m_waiter_head) {
+        cond->m_waiter_tail = (void *)&waiter;
+    } else {
+        ((cond_waiter_t *)cond->m_waiter_head)->m_prev = (void *)&waiter;
+    }
+    cond->m_waiter_head = (void *)&waiter;
+    opal_atomic_unlock(&cond->m_lock);
+
+    while (1) {
+        opal_mutex_unlock(lock);
+        qthread_yield();
+        opal_mutex_lock(lock);
+        /* Check if someone woke me up. */
+        opal_atomic_lock(&cond->m_lock);
+        int signaled = waiter.m_signaled;
+        opal_atomic_unlock(&cond->m_lock);
```

Comment:
This lock (or any memory barrier that can order accesses to `m_signaled`) is necessary.
`m_signaled` can be updated by another thread that runs on another worker (=Pthreads), but in my understanding, the C standard does not guarantee when a reader can see `m_signaled` updated by a writer without atomic operations.

Since this `waiter` object is allocated in a function stack, it can cause SEGV if the writer accesses this object after `m_signaled` is read by a reader.  It should not happen in `opal_cond_signal()` if all the memory write operations are issued in the order written in the code.

If acquire-release or seq-cst load/store can remove this lock, but it makes the code a bit more complicated.

Reply:
Ahh I see what you mean. The problem is that you are issuing a lot of atomic memory operations, esp. if `qthread_yield` is a no-op (i.e., if there are no other qthreads available for execution). In that short snippet, it's four atomic memory ops already. If you have multiple threads (not qthreads) waiting on this conditional variable these locks will be quite contended. Plus, `opal_mutex_unlock` calls `qthread_yield` again (which I am also not convinced is necessary).

Anyway, what is the `lock` argument in this function good for? It doesn't seem to protect anything so it can be re-taken right before exiting the function (the `cond` has its own lock). If you make `waiter.m_signaled` `volatile` (to prevent the compiler from optimizing away the loads) and insert a memory barrier (`opal_atomic_wmb()` should do) in the signalling code before setting `m_signaled` then the tail of the function could look like this:

```C
    while (1) {
        qthread_yield();
        /* Check if someone woke me up. */
        if (waiter.m_signaled) {
            break;
        }
    }
    opal_mutex_lock(lock);
    return OPAL_SUCCESS;
```

Yes, the signalling will become a bit more expensive due to the memory barrier but having several threads hammering on atomic locks is by far more problematic than a few more cycles spent by one thread in the signalling code.

Reply:
Indeed this type of polling is probably not the best use of Qthreads.  Using FEBs instead would make for an event-driven approach.  Not sure if its best to leave it for now and reimplement later or go ahead and fix now. @janciesko